internal/tls/tls.go

```diff
@@ -139,12 +146,26 @@ func NewServerConfig(opts ...Option) (*Config, error) {
 		c.sn = "vald-server"
 		c.cfg.ServerName = c.sn
 	}
-	// load cert pair
-	kp, err := loadKeyPair(c.sn, c.cert, c.key)
-	if err != nil {
-		return nil, err
+	// Configure certificate strategy.
+	if c.hotReload {
+		// Do not preload Certificates; use GetCertificate instead.
+		c.cfg.GetCertificate = func(chi *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			kp, err := loadKeyPair(c.sn, c.cert, c.key)
+			if err != nil {
+				return nil, err
+			}
+			c.store.Set(&kp)
+			return c.store.Get(), nil
+		}
+	} else {
+		// load once statically
+		kp, err := loadKeyPair(c.sn, c.cert, c.key)
+		if err != nil {
+			return nil, err
+		}
+		c.cfg.Certificates = []tls.Certificate{kp}
+		c.store.Set(&kp)
 	}
```

Comment:
When using GetCertificate, cfg.Certificates should still be set for NameToCertificate. So this conditional branch only sets one or the other, which causes problems.